// @badeball/cypress-cucumber-preprocessor/dist/plugin-event-handlers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStringAttachmentHandler = exports.testCaseFinishedHandler = exports.testStepFinishedHandler = exports.testStepStartedHandler = exports.testCaseStartedHandler = exports.specEnvelopesHandler = exports.afterScreenshotHandler = exports.afterSpecHandler = exports.beforeSpecHandler = exports.afterRunHandler = exports.beforeRunHandler = void 0;
const fs_1 = __importStar(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const promises_1 = require("stream/promises");
const stream_1 = __importDefault(require("stream"));
const chalk_1 = __importDefault(require("chalk"));
const messages = __importStar(require("@cucumber/messages"));
const ci_environment_1 = __importDefault(require("@cucumber/ci-environment"));
const split_1 = __importDefault(require("split"));
const constants_1 = require("./constants");
const preprocessor_configuration_1 = require("./preprocessor-configuration");
const paths_1 = require("./helpers/paths");
const messages_1 = require("./helpers/messages");
const memoize_1 = require("./helpers/memoize");
const debug_1 = __importDefault(require("./helpers/debug"));
const error_1 = require("./helpers/error");
const assertions_1 = require("./helpers/assertions");
const formatters_1 = require("./helpers/formatters");
const colors_1 = require("./helpers/colors");
const type_guards_1 = require("./helpers/type-guards");
const version_1 = require("./version");
const resolve = (0, memoize_1.memoize)(preprocessor_configuration_1.resolve);
let state = {
    state: "uninitialized",
};
const isFeature = (spec) => spec.name.endsWith(".feature");
const end = (stream) => new Promise((resolve) => stream.end(resolve));
const createPrettyStream = () => {
    const line = (0, split_1.default)();
    const indent = new stream_1.default.Transform({
        objectMode: true,
        transform(chunk, _, callback) {
            callback(null, chunk.length === 0 ? "" : "  " + chunk);
        },
    });
    const log = new stream_1.default.Writable({
        write(chunk, _, callback) {
            console.log(chunk.toString("utf8"));
            callback();
        },
    });
    return stream_1.default.compose(line, indent, log);
};
const createStateError = (stateHandler, currentState) => new error_1.CypressCucumberError(`Unexpected state in ${stateHandler}: ${currentState}. This almost always means that you or some other plugin, are overwriting this plugin's event handlers. For more information & workarounds, see https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/docs/event-handlers.md (if neither workaround work, please report at ${error_1.homepage})`);
async function beforeRunHandler(config) {
    (0, debug_1.default)("beforeRunHandler()");
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return;
    }
    switch (state.state) {
        case "uninitialized":
            break;
        default:
            throw (0, error_1.createError)("Unexpected state in beforeRunHandler: " + state.state);
    }
    // Copied from https://github.com/cucumber/cucumber-js/blob/v10.0.1/src/cli/helpers.ts#L104-L122.
    const meta = {
        meta: {
            protocolVersion: messages.version,
            implementation: {
                version: version_1.version,
                name: "@badeball/cypress-cucumber-preprocessor",
            },
            cpu: {
                name: os_1.default.arch(),
            },
            os: {
                name: os_1.default.platform(),
                version: os_1.default.release(),
            },
            runtime: {
                name: "node.js",
                version: process.versions.node,
            },
            ci: (0, ci_environment_1.default)(process.env),
        },
    };
    const testRunStarted = {
        testRunStarted: {
            timestamp: (0, messages_1.createTimestamp)(),
        },
    };
    state = {
        state: "before-run",
        messages: {
            accumulation: [meta, testRunStarted],
        },
    };
}
exports.beforeRunHandler = beforeRunHandler;
async function afterRunHandler(config) {
    (0, debug_1.default)("afterRunHandler()");
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return;
    }
    switch (state.state) {
        case "after-spec": // This is the normal case.
        case "before-run": // This can happen when running only non-feature specs.
            break;
        default:
            throw (0, error_1.createError)("Unexpected state in afterRunHandler: " + state.state);
    }
    const testRunFinished = {
        testRunFinished: {
            /**
             * We're missing a "success" attribute here, but cucumber-js doesn't output it, so I won't.
             * Mostly because I don't want to look into the semantics of it right now.
             */
            timestamp: (0, messages_1.createTimestamp)(),
        },
    };
    state = {
        state: "after-run",
        messages: {
            accumulation: state.messages.accumulation.concat(testRunFinished),
        },
    };
    if (preprocessor.messages.enabled) {
        const messagesPath = (0, paths_1.ensureIsAbsolute)(config.projectRoot, preprocessor.messages.output);
        await fs_1.promises.mkdir(path_1.default.dirname(messagesPath), { recursive: true });
        await fs_1.promises.writeFile(messagesPath, state.messages.accumulation
            .map((message) => JSON.stringify(message))
            .join("\n") + "\n");
    }
    if (preprocessor.json.enabled) {
        const jsonPath = (0, paths_1.ensureIsAbsolute)(config.projectRoot, preprocessor.json.output);
        await fs_1.promises.mkdir(path_1.default.dirname(jsonPath), { recursive: true });
        let jsonOutput;
        const eventBroadcaster = (0, formatters_1.createJsonFormatter)(state.messages.accumulation, (chunk) => {
            jsonOutput = chunk;
        });
        for (const message of state.messages.accumulation) {
            eventBroadcaster.emit("envelope", message);
        }
        (0, assertions_1.assertIsString)(jsonOutput, "Expected JSON formatter to have finished, but it never returned");
        await fs_1.promises.writeFile(jsonPath, jsonOutput);
    }
    if (preprocessor.html.enabled) {
        const htmlPath = (0, paths_1.ensureIsAbsolute)(config.projectRoot, preprocessor.html.output);
        await fs_1.promises.mkdir(path_1.default.dirname(htmlPath), { recursive: true });
        const output = fs_1.default.createWriteStream(htmlPath);
        await (0, promises_1.pipeline)(stream_1.default.Readable.from(state.messages.accumulation), (0, formatters_1.createHtmlStream)(), output);
    }
}
exports.afterRunHandler = afterRunHandler;
async function beforeSpecHandler(config, spec) {
    (0, debug_1.default)("beforeSpecHandler()");
    if (!isFeature(spec)) {
        return;
    }
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return;
    }
    switch (state.state) {
        case "before-run":
        case "after-spec":
            {
                if (preprocessor.pretty.enabled) {
                    const writable = createPrettyStream();
                    const eventBroadcaster = (0, formatters_1.createPrettyFormatter)((0, colors_1.useColors)(), (chunk) => writable.write(chunk));
                    state = {
                        state: "before-spec",
                        pretty: {
                            enabled: true,
                            broadcaster: eventBroadcaster,
                            writable,
                        },
                        messages: state.messages,
                    };
                }
                else {
                    state = {
                        state: "before-spec",
                        pretty: {
                            enabled: false,
                        },
                        messages: state.messages,
                    };
                }
            }
            break;
        // This happens in case of visting a new domain, ref. https://github.com/cypress-io/cypress/issues/26300.
        // In this case, we want to disgard messages obtained in the current test and allow execution to continue
        // as if nothing happened.
        case "before-spec":
        case "step-started":
            break;
        default:
            throw createStateError("beforeSpecHandler", state.state);
    }
}
exports.beforeSpecHandler = beforeSpecHandler;
async function afterSpecHandler(config, spec, results) {
    (0, debug_1.default)("afterSpecHandler()");
    if (!isFeature(spec)) {
        return;
    }
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return;
    }
    switch (state.state) {
        case "test-finished": // This is the normal case.
        case "before-spec": // This can happen if a spec doesn't contain any tests.
        case "received-envelopes": // This can happen in case of a failing beforeEach hook.
            break;
        default:
            throw (0, error_1.createError)("Unexpected state in afterSpecHandler: " + state.state);
    }
    if ("pretty" in state && state.pretty.enabled) {
        await end(state.pretty.writable);
    }
    // `results` is undefined when running via `cypress open`.
    // However, `isTrackingState` is never true in open-mode, thus this should be defined.
    (0, assertions_1.assert)(results, "Expected results to be defined");
    const wasRemainingSkipped = results.tests.some((test) => { var _a; return (_a = test.displayError) === null || _a === void 0 ? void 0 : _a.match(constants_1.HOOK_FAILURE_EXPR); });
    if (wasRemainingSkipped) {
        console.log(chalk_1.default.yellow(`  Hook failures can't be represented in any reports (messages / json / html), thus none is created for ${spec.relative}.`));
        state = {
            state: "after-spec",
            messages: {
                accumulation: state.messages.accumulation,
            },
        };
    }
    else {
        // IE. the spec didn't contain any tests.
        if (state.state === "before-spec") {
            state = {
                state: "after-spec",
                messages: {
                    accumulation: state.messages.accumulation,
                },
            };
        }
        else {
            state = {
                state: "after-spec",
                messages: {
                    accumulation: state.messages.accumulation.concat(state.messages.current),
                },
            };
        }
    }
}
exports.afterSpecHandler = afterSpecHandler;
async function afterScreenshotHandler(config, details) {
    (0, debug_1.default)("afterScreenshotHandler()");
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return details;
    }
    switch (state.state) {
        case "step-started":
            break;
        default:
            return details;
    }
    let buffer;
    try {
        buffer = await fs_1.promises.readFile(details.path);
    }
    catch (_a) {
        return details;
    }
    const message = {
        attachment: {
            testCaseStartedId: state.testCaseStartedId,
            testStepId: state.testStepStartedId,
            body: buffer.toString("base64"),
            mediaType: "image/png",
            contentEncoding: "BASE64",
        },
    };
    state.messages.current.push(message);
    return details;
}
exports.afterScreenshotHandler = afterScreenshotHandler;
async function specEnvelopesHandler(config, data) {
    (0, debug_1.default)("specEnvelopesHandler()");
    switch (state.state) {
        case "before-spec":
            break;
        // This happens in case of visting a new domain, ref. https://github.com/cypress-io/cypress/issues/26300.
        // In this case, we want to disgard messages obtained in the current test and allow execution to continue
        // as if nothing happened.
        case "step-started":
            {
                const iTestCaseStarted = state.messages.current.findLastIndex((message) => !!message.testCaseStarted);
                if (iTestCaseStarted === -1) {
                    throw (0, error_1.createError)("Expected to find a testCaseStarted envelope");
                }
                let pretty;
                if (state.pretty.enabled) {
                    await end(state.pretty.writable);
                    console.log("  Reloading..");
                    console.log();
                    const writable = createPrettyStream();
                    const eventBroadcaster = (0, formatters_1.createPrettyFormatter)((0, colors_1.useColors)(), (chunk) => writable.write(chunk));
                    for (const message of data.messages) {
                        eventBroadcaster.emit("envelope", message);
                    }
                    pretty = {
                        enabled: true,
                        writable,
                        broadcaster: eventBroadcaster,
                    };
                }
                else {
                    pretty = state.pretty;
                }
                state = {
                    state: "received-envelopes",
                    pretty,
                    messages: {
                        accumulation: state.messages.accumulation,
                        current: state.messages.current.slice(0, iTestCaseStarted),
                    },
                };
            }
            return true;
        default:
            throw createStateError("specEnvelopesHandler", state.state);
    }
    if (state.pretty.enabled) {
        for (const message of data.messages) {
            state.pretty.broadcaster.emit("envelope", message);
        }
    }
    state = {
        state: "received-envelopes",
        pretty: state.pretty,
        messages: {
            accumulation: state.messages.accumulation,
            current: data.messages,
        },
    };
    return true;
}
exports.specEnvelopesHandler = specEnvelopesHandler;
function testCaseStartedHandler(config, data) {
    (0, debug_1.default)("testCaseStartedHandler()");
    switch (state.state) {
        case "received-envelopes":
        case "test-finished":
            break;
        default:
            throw createStateError("testCaseStartedHandler", state.state);
    }
    if (state.pretty.enabled) {
        state.pretty.broadcaster.emit("envelope", {
            testCaseStarted: data,
        });
    }
    state = {
        state: "test-started",
        pretty: state.pretty,
        messages: {
            accumulation: state.messages.accumulation,
            current: state.messages.current.concat({ testCaseStarted: data }),
        },
        testCaseStartedId: data.id,
    };
    return true;
}
exports.testCaseStartedHandler = testCaseStartedHandler;
function testStepStartedHandler(config, data) {
    (0, debug_1.default)("testStepStartedHandler()");
    switch (state.state) {
        case "test-started":
        case "step-finished":
            break;
        // This state can happen in cases where an error is "rescued".
        case "step-started":
            break;
        default:
            throw createStateError("testStepStartedHandler", state.state);
    }
    if (state.pretty.enabled) {
        state.pretty.broadcaster.emit("envelope", {
            testStepStarted: data,
        });
    }
    state = {
        state: "step-started",
        pretty: state.pretty,
        messages: {
            accumulation: state.messages.accumulation,
            current: state.messages.current.concat({ testStepStarted: data }),
        },
        testCaseStartedId: state.testCaseStartedId,
        testStepStartedId: data.testStepId,
    };
    return true;
}
exports.testStepStartedHandler = testStepStartedHandler;
async function testStepFinishedHandler(config, options, testStepFinished) {
    var _a;
    (0, debug_1.default)("testStepFinishedHandler()");
    switch (state.state) {
        case "step-started":
            break;
        default:
            throw createStateError("testStepFinishedHandler", state.state);
    }
    if (state.pretty.enabled) {
        state.pretty.broadcaster.emit("envelope", {
            testStepFinished,
        });
    }
    const { testCaseStartedId, testStepId } = testStepFinished;
    const { testCaseId: pickleId } = (0, assertions_1.assertAndReturn)(state.messages.current
        .map((message) => message.testCaseStarted)
        .filter(type_guards_1.notNull)
        .find((testCaseStarted) => testCaseStarted.id === testCaseStartedId), "Expected to find a testCaseStarted");
    const testCase = (0, assertions_1.assertAndReturn)(state.messages.current
        .map((message) => message.testCase)
        .filter(type_guards_1.notNull)
        .find((testCase) => testCase.id === pickleId), "Expected to find a testCase");
    const { pickleStepId, hookId } = (0, assertions_1.assertAndReturn)(testCase.testSteps.find((testStep) => testStep.id === testStepId), "Expected to find a testStep");
    if (pickleStepId != null) {
        const pickle = (0, assertions_1.assertAndReturn)(state.messages.current
            .map((message) => message.pickle)
            .filter(type_guards_1.notNull)
            .find((pickle) => pickle.id === pickleId), "Expected to find a pickle");
        const pickleStep = (0, assertions_1.assertAndReturn)(pickle.steps.find((step) => step.id === pickleStepId), "Expected to find a pickleStep");
        const gherkinDocument = (0, assertions_1.assertAndReturn)(state.messages.current
            .map((message) => message.gherkinDocument)
            .filter(type_guards_1.notNull)
            .find((gherkinDocument) => gherkinDocument.uri === pickle.uri), "Expected to find a gherkinDocument");
        const attachments = [];
        await ((_a = options.onAfterStep) === null || _a === void 0 ? void 0 : _a.call(options, {
            result: testStepFinished.testStepResult,
            pickle,
            pickleStep,
            gherkinDocument,
            testCaseStartedId,
            testStepId,
            attach(data, mediaType) {
                if (typeof data === "string") {
                    mediaType = mediaType !== null && mediaType !== void 0 ? mediaType : "text/plain";
                    if (mediaType.startsWith("base64:")) {
                        attachments.push({
                            data,
                            mediaType: mediaType.replace("base64:", ""),
                            encoding: messages.AttachmentContentEncoding.BASE64,
                        });
                    }
                    else {
                        attachments.push({
                            data,
                            mediaType: mediaType !== null && mediaType !== void 0 ? mediaType : "text/plain",
                            encoding: messages.AttachmentContentEncoding.IDENTITY,
                        });
                    }
                }
                else if (data instanceof Buffer) {
                    if (typeof mediaType !== "string") {
                        throw Error("Buffer attachments must specify a media type");
                    }
                    attachments.push({
                        data: data.toString("base64"),
                        mediaType,
                        encoding: messages.AttachmentContentEncoding.BASE64,
                    });
                }
                else {
                    throw Error("Invalid attachment data: must be a Buffer or string");
                }
            },
        }));
        for (const attachment of attachments) {
            await createStringAttachmentHandler(config, attachment);
        }
    }
    else {
        (0, assertions_1.assert)(hookId != null, "Expected a hookId in absence of pickleStepId");
    }
    state = {
        state: "step-finished",
        pretty: state.pretty,
        messages: {
            accumulation: state.messages.accumulation,
            current: state.messages.current.concat({ testStepFinished }),
        },
        testCaseStartedId: state.testCaseStartedId,
    };
    return true;
}
exports.testStepFinishedHandler = testStepFinishedHandler;
function testCaseFinishedHandler(config, data) {
    (0, debug_1.default)("testCaseFinishedHandler()");
    switch (state.state) {
        case "test-started":
        case "step-finished":
            break;
        default:
            throw createStateError("testCaseFinishedHandler", state.state);
    }
    if (state.pretty.enabled) {
        state.pretty.broadcaster.emit("envelope", {
            testCaseFinished: data,
        });
    }
    state = {
        state: "test-finished",
        pretty: state.pretty,
        messages: {
            accumulation: state.messages.accumulation,
            current: state.messages.current.concat({ testCaseFinished: data }),
        },
    };
    return true;
}
exports.testCaseFinishedHandler = testCaseFinishedHandler;
async function createStringAttachmentHandler(config, { data, mediaType, encoding }) {
    (0, debug_1.default)("createStringAttachmentHandler()");
    const preprocessor = await resolve(config, config.env, "/");
    if (!preprocessor.isTrackingState) {
        return true;
    }
    switch (state.state) {
        case "step-started":
            break;
        default:
            throw createStateError("createStringAttachmentHandler", state.state);
    }
    const message = {
        attachment: {
            testCaseStartedId: state.testCaseStartedId,
            testStepId: state.testStepStartedId,
            body: data,
            mediaType: mediaType,
            contentEncoding: encoding,
        },
    };
    state.messages.current.push(message);
    return true;
}
exports.createStringAttachmentHandler = createStringAttachmentHandler;
